data_model/src/name.rs

```diff
@@ -91,9 +100,9 @@ impl FromStr for Name {
 ///
 /// All of the given pointers must be valid
 #[no_mangle]
-#[cfg(feature = "ffi")]
+#[cfg(feature = "ffi_api")]
 #[allow(non_snake_case, unsafe_code)]
-pub unsafe extern "C" fn Name__from_str<'itm>(
+unsafe extern "C" fn Name__from_str<'itm>(
```

Comment:
as far as I understand extern functions will be visible to the linker even if private. This way the public api isn't affected

Reply:
> We should also check with many linkers. At the very least we need ld, lld and mold.

this is not related to linkers. It's that `rustc` will not remove extern functions, even if they are private, when building libraries. LTO on a `cdylib` might affect this behavior. I can't find the issue discussion about it now, but this cannot go unseen so I'll definitely address this in the future.

Reply:
> This If should be turned into a test. Trybuild is normally there for macros, but we can use it to check if this really affects the linker.

ok, but why test compiler behavior?

> We should also check with many linkers. At the very least we need `ld`, `lld` and `mold`.

IMO this is about `rustc` not removing `extern` functions, even if they are private. `LTO` on a `cdylib` might affect this behavior. I can't find the issue discussion about it now, but this cannot go unseen so I'll definitely address this in the future. I've opened an issue

Reply:
> ok, but why test compiler behavior?

Because in this instance we're relying on it. It's not 100% clear what to do with a private `extern` function. `rustc` might keep them, might change its mind to optimise binary size, and `mold` might start doing some thin LTO by default, which we would notice by a test failing. 

The preferred alternative is, of course, to enforce the inclusion by making the function public.

Reply:
> The preferred alternative is, of course, to enforce the inclusion by making the function public.

restored it to public visibility